### ISSUE TITLE
Consolidate Projection for Schema and RecordBatch

### ISSUE THREAD

**Issue by alamb** [apache/arrow-datafusion#1425]
**Background**

@hntd187  fixed https://github.com/apache/arrow-datafusion/issues/1361 via https://github.com/apache/arrow-datafusion/pull/1378 but when I was reviewing the code, I found several other places that project `RecordBatch`s and `Schemas` that may also have the same subtle issues about losing the metadata. I am not sure of any bugs related to this yet but I fear they are lurking

The basic idea is to make functions like the following (which handle metadata correctly, following the pattern in #1361 )

```rust
fn project_schema(schema: &Schema, projection: &[usize]) -> <Schema> {
...
}

fn project_batch(batch: &RecordBatch, projection: &[usize]) -> Result<RecordBatch> {
...
}
```
And replace the duplicated code like

```rust
        let projected_schema = match &projection {
            Some(columns) => {
                let fields: Result<Vec<Field>> = columns
                    .iter()
                    .map(|i| {
                        if *i < schema.fields().len() {
                            Ok(schema.field(*i).clone())
                        } else {
                            Err(DataFusionError::Internal(
                                "Projection index out of range".to_string(),
                            ))
                        }
                    })
                    .collect();
                Arc::new(Schema::new(fields?))
            }
            None => Arc::clone(&schema),
        };
```

And 

```rust
                Some(columns) => Some(RecordBatch::try_new(
                    self.schema.clone(),
                    columns.iter().map(|i| batch.column(*i).clone()).collect(),
                )),

```

ALl over the datafusion codebase

**Additional context**
Here is a corresponding arrow ticket to put the logic into arrow-rs: https://github.com/apache/arrow-rs/issues/1014

**Comment by alamb** [comment]
The code for `Schema::project` and `RecordBatch::project` from @hntd187 has been merged into https://github.com/apache/arrow-rs/pull/1033

Once that is available in a release (likely `arrow 6.5.0` next week) we can clean up some of this code in DataFusion

**Comment by alamb** [comment]
Closed by https://github.com/apache/arrow-datafusion/pull/1638
